### ISSUE TITLE
Unresolved reference in expo-permissions BackgroundLocationRequester.kt

### ISSUE THREAD

**Issue by Sheharyar566** [expo/expo#12590]
I am working on an ejected workflow and was trying to remove the background location permissions using app.json android/permissions key-pair. Before doing that, both the build and the app was working fine. After I added them as
`
"permissions":[ "SYSTEM_ALERT_WINDOW", "READ_PHONE_STATE", "USE_FINGERPRINT", "USE_BIOMETRIC", "VIBRATE", "MODIFY_AUDIO_SETTINGS", "ACCESS_COARSE_LOCATION", "ACCESS_FINE_LOCATION", "CAMERA", "RECORD_AUDIO", "READ_EXTERNAL_STORAGE", "WRITE_EXTERNAL_STORAGE", "WRITE_SETTINGS" ]
`
and removed the node_modules folder as well as the package-lock.json file. After this i ran npm install and then rebuild project using android studio, however I get this error:

`
> Task :expo-permissions:compileDebugKotlin FAILED
e: E:\Work\pax-app\node_modules\expo-image-picker\node_modules\expo-permissions\android\src\main\java\expo\modules\permissions\requesters\BackgroundLocationRequester.kt: (12, 50): Unresolved reference: R
e: E:\Work\pax-app\node_modules\expo-image-picker\node_modules\expo-permissions\android\src\main\java\expo\modules\permissions\requesters\BackgroundLocationRequester.kt: (19, 52): Unresolved reference: R
e: E:\Work\pax-app\node_modules\expo-image-picker\node_modules\expo-permissions\android\src\main\java\expo\modules\permissions\requesters\BackgroundLocationRequester.kt: (40, 36): Unresolved reference: R
`

### Do you understand that any discussions or questions opened as issues in the core Expo repository will be closed?

Yes

**Comment by byCedric** [comment]
Hi @Sheharyar566! It looks like it's failing on [this line](https://github.com/expo/expo/blob/master/packages/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt#L12), which means [`Build.VERSION_CODES.R`](https://developer.android.com/reference/android/os/Build.VERSION_CODES#R) is not available in your native code.

I think there are 3 options for you to try:
1. Update the compile API to 30, which should include that build code.
2. Replace the reference with `Constant Value: 30 (0x0000001e)` using something like [patch-package](https://www.npmjs.com/package/patch-package)
3. If you are using `expo-location`, you might have more luck using the permission requesters from that package.

We [marked the `expo-permission` as deprecated since SDK 41](https://blog.expo.io/expo-sdk-41-12cc5232f2ef), I would recommend trying out option 3 first. Hope this helps!

**Comment by Sheharyar566** [comment]
Hi @byCedric , really thankful for the the list of available options. However, it seems like I forgot to mention that I'm at SDK v 39. So should it really act like that for SDK 39?

**Comment by TheoBr** [comment]
Running into this as well. Appears that `expo-permissions` is a dependency of react-native-unimodules

![image](https://user-images.githubusercontent.com/6751787/123029409-f04b8100-d395-11eb-8dad-aeb6ab3e3703.png)

**Comment by brentvatne** [comment]
you can exclude it if you want. see "Need to exclude some unimodules that are being automatically linked?"

**Comment by TheoBr** [comment]
@brentvatne Good call, will filter these out in our case.

My first pass was the patch-package route and it went well. I think `Build.VERSION_CODES.R` only exists on buildTools SDK version 30, but compatibility is fine otherwise? IMO, might be worth not using that particular build variable as it is not included on v29

(We're on `v29.0.2`, entirely possible that it was added on `v29.0.3`, have not checked)

**Comment by RodolfoGS** [comment]
I created the patch if someone need it.

Filename:
`expo-permissions+12.1.1.patch`

Content:
```
diff --git a/node_modules/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt b/node_modules/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt
index a78d902..13510d4 100644
--- a/node_modules/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt
+++ b/node_modules/expo-permissions/android/src/main/java/expo/modules/permissions/requesters/BackgroundLocationRequester.kt
@@ -9,19 +9,19 @@ import expo.modules.interfaces.permissions.PermissionsStatus
 
 class BackgroundLocationRequester : PermissionRequester {
   override fun parseAndroidPermissions(permissionsResponse: Map<String, PermissionsResponse>) = when {
-    Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> parseAndroidPermissionsForAndroidR(permissionsResponse)
-    Build.VERSION.SDK_INT == Build.VERSION_CODES.Q -> parseAndroidPermissionsForAndroidQ(permissionsResponse)
+    Build.VERSION.SDK_INT >= 30 -> parseAndroidPermissionsForAndroidR(permissionsResponse)
+    Build.VERSION.SDK_INT == 29 -> parseAndroidPermissionsForAndroidQ(permissionsResponse)
     else -> parseAndroidPermissionsForLegacyAndroids(permissionsResponse)
   }
 
   override fun getAndroidPermissions(): List<String> {
     return when {
-      Build.VERSION.SDK_INT >= Build.VERSION_CODES.R -> {
+      Build.VERSION.SDK_INT >= 30 -> {
         listOf(
           Manifest.permission.ACCESS_BACKGROUND_LOCATION
         )
       }
-      Build.VERSION.SDK_INT == Build.VERSION_CODES.Q -> {
+      Build.VERSION.SDK_INT == 29 -> {
         listOf(
           Manifest.permission.ACCESS_FINE_LOCATION,
           Manifest.permission.ACCESS_COARSE_LOCATION,
@@ -37,7 +37,7 @@ class BackgroundLocationRequester : PermissionRequester {
     }
   }
 
-  @RequiresApi(Build.VERSION_CODES.R)
+  @RequiresApi(30)
   private fun parseAndroidPermissionsForAndroidR(permissionsResponse: Map<String, PermissionsResponse>): Bundle {
     return Bundle().apply {
       val backgroundLocation = permissionsResponse.getValue(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
@@ -51,7 +51,7 @@ class BackgroundLocationRequester : PermissionRequester {
     }
   }
 
-  @RequiresApi(Build.VERSION_CODES.Q)
+  @RequiresApi(29)
   private fun parseAndroidPermissionsForAndroidQ(permissionsResponse: Map<String, PermissionsResponse>): Bundle {
     return Bundle().apply {
       val accessFineLocation = permissionsResponse.getValue(Manifest.permission.ACCESS_FINE_LOCATION)
```

**Comment by brentvatne** [comment]
this shouldn't be a problem on sdk 42. expo-permissions is deprecated, no longer documented, and we suggest switching away to the module specific permission requesters like `Camera.requestPermissionsAsync()`

**Comment by Un3qual** [comment]
It looks like `expo-image-picker` still depends on `expo-permissions` even on master. Will this be removed in the future?
https://github.com/expo/expo/blob/ff1e3b9e47b4c6b93d898a22bf394aeb428713fd/packages/expo-image-picker/package.json#L41

**Comment by brentvatne** [comment]
that is a mistake. it is unused by the library but included as a dependency. we'll resolve that

**Comment by brentvatne** [comment]
@Un3qual - fixed in expo-image-picker@10.2.3
